src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,747 @@
+//   - Setting up callbacks for when the number of windows changes.
+//   - Setting up the throttled func for layout persistence. Arguments:
+// - <none>
+void WindowEmperor::_becomeMonarch()
+{
+    // Add a callback to the window manager so that when the Monarch wants a new
+    // window made, they come to us
+    _manager.RequestNewWindow([this](auto&&, const Remoting::WindowRequestedArgs& args) {
+        _createNewWindowThread(args);
+    });
+
+    _createMessageWindow();
+
+    _setupGlobalHotkeys();
+
+    // When the settings change, we'll want to update our global hotkeys and our
+    // notification icon based on the new settings.
+    _app.Logic().SettingsChanged([this](auto&&, const TerminalApp::SettingsLoadEventArgs& args) {
+        if (SUCCEEDED(args.Result()))
+        {
+            _setupGlobalHotkeys();
+            _checkWindowsForNotificationIcon();
+        }
+    });
+
+    // On startup, immediately check if we need to show the notification icon.
+    _checkWindowsForNotificationIcon();
+
+    // Set the number of open windows (so we know if we are the last window)
+    // and subscribe for updates if there are any changes to that number.
+
+    _revokers.WindowCreated = _manager.WindowCreated(winrt::auto_revoke, { this, &WindowEmperor::_numberOfWindowsChanged });
+    _revokers.WindowClosed = _manager.WindowClosed(winrt::auto_revoke, { this, &WindowEmperor::_numberOfWindowsChanged });
+
+    // If the monarch receives a QuitAll event it will signal this event to be
+    // ran before each peasant is closed.
+    _revokers.QuitAllRequested = _manager.QuitAllRequested(winrt::auto_revoke, { this, &WindowEmperor::_quitAllRequested });
+
+    // The monarch should be monitoring if it should save the window layout.
+    // We want at least some delay to prevent the first save from overwriting
+    _getWindowLayoutThrottler.emplace(std::move(std::chrono::seconds(10)), std::move([this]() { _saveWindowLayoutsRepeat(); }));
+    _getWindowLayoutThrottler.value()();
+
+    // BODGY
+    //
+    // We've got a weird crash that happens terribly inconsistently, but pretty
+    // readily on migrie's laptop, only in Debug mode. Apparently, there's some
+    // weird ref-counting magic that goes on during teardown, and our
+    // Application doesn't get closed quite right, which can cause us to crash
+    // into the debugger. This of course, only happens on exit, and happens
+    // somewhere in the XamlHost.dll code.
+    //
+    // Crazily, if we _manually leak the Application_ here, then the crash
+    // doesn't happen. This doesn't matter, because we really want the
+    // Application to live for _the entire lifetime of the process_, so the only
+    // time when this object would actually need to get cleaned up is _during
+    // exit_. So we can safely leak this Application object, and have it just
+    // get cleaned up normally when our process exits.
+    auto a{ _app };
+    ::winrt::detach_abi(a);

```

Comment:
Oof... It's probably rather indicative of our many many potential sites of memory corruption. 😄 It's of course rather difficult to find an example right now to make a point, but it's possible to find a number of contrieved examples at least. For instance it's possible for `WindowManager` to outlive the `WindowEmperor`, because `WindowThreads` holds a weak ref to `WindowManager` and a strong ref to `WindowEmperor`. So it's possible that sometimes, depending on a race condition during shutdown when creating windows at the same time, that the `RequestNewWindow` callback is invoked and accesses deallocated memory because it's referencing `this`. This access will likely succeed because `malloc()` and `new` return previously used memory and the memory corruption will then occur in `XamlHost.dll` for instance. I'd rather see this issue fixed before merging, but I'd understand if you don't want to pursue it just yet. I suspect however that this bug will also occur in Release builds by crashing during long sessions (memory memory corruptions more apparent).

Reply:
refer also to https://github.com/microsoft/terminal/pull/14843#discussion_r1131414165

this is code we used to have because _b o d g e_, and never understood why

Reply:
Well that's pretty scary then. 😄